testproject/testapp/tests/social/test_provider_auth.py

```diff
@@ -76,7 +98,127 @@ def test_post_facebook_provider_validation_fails_if_invalid_state(self):
             return_value=data["state"][::-1],
         ).start()
 
-        request = self.factory.post()
+        request = self.factory.post("/auth/facebook/")
+        request.GET = {k: v for k, v in data.items()}
+        response = self._get_view_response(request, provider="facebook")
+        assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+    def test_post_facebook_provider_auth_forbidden_error(self):
+        """Test handling of AuthForbidden exception."""
+        data = {"code": "XYZ", "state": "ABC"}
+
+        mock.patch(
+            "social_core.backends.facebook.FacebookOAuth2.auth_complete",
+            side_effect=AuthForbidden(backend=None),
+        ).start()
+        mock.patch(
+            "social_core.backends.oauth.OAuthAuth.get_session_state",
+            return_value=data["state"],
+        ).start()
+
+        request = self.factory.post("/auth/facebook/")
+        request.GET = {k: v for k, v in data.items()}
+        response = self._get_view_response(request, provider="facebook")
+        assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+    def test_post_facebook_provider_auth_canceled_error(self):
+        """Test handling of AuthCanceled exception."""
+        data = {"code": "XYZ", "state": "ABC"}
+
+        mock.patch(
+            "social_core.backends.facebook.FacebookOAuth2.auth_complete",
+            side_effect=AuthCanceled(backend=None),
+        ).start()
+        mock.patch(
+            "social_core.backends.oauth.OAuthAuth.get_session_state",
+            return_value=data["state"],
+        ).start()
+
+        request = self.factory.post("/auth/facebook/")
+        request.GET = {k: v for k, v in data.items()}
+        response = self._get_view_response(request, provider="facebook")
+        assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+    def test_post_facebook_provider_auth_unknown_error(self):
```

Comment:
looks like the code for all 3 tests is the same, I would do a test generator or anything to decrease the amount of test code to read / maintain